src/performance.js

```diff
@@ -0,0 +1,147 @@
+ * along with the Game Closure SDK.  If not, see <http://mozilla.org/MPL/2.0/>.
+ */
+
+ /**
+ * @module performanceMetric
+ *
+ * Measures performance by keeping a window of delta times and providing a score using that window.
+ *
+**/
+
+var engine = null;
+
+var DEFAULT_RANK = 0;
+var DEFAULT_ALLOW_REDUCTION = true;
+var LOWER_BOUND = 10;
+var HIGHER_BOUND = 60;
+var TICKS_TIL_INCREASE_SCORE = 50;
+  
+var Performance = Class(function () {
+  var ticksSinceLastWorstUpdate = 0;
+  var worstTicksSize = 30;
+  var worstTicks = [];
+  var minFPS = LOWER_BOUND;
+  var maxFPS = HIGHER_BOUND;
+  var lastTick;
+
+  var _map = function(val, start1, stop1, start2, stop2) {
+    var range1 = stop1 - start1;
+    var range2 = stop2 - start2;
+    var valPosition = val - start1;
+    var valuePercentage = valPosition / range1;
+    return start2 + range2 * valuePercentage;
+  };
+
+  this.init = function () {
+    this.measuring = false;
+  };
+
+  this.setTargetFPSRange = function (min, max) {
+    minFPS = min;
+    maxFPS = max;
+  };
+
+  this.startMeasuring = function() {
+    if (!this.measuring) {
+      this.measuring = true;
+      if (engine === null) {
+        import ui.Engine as Engine;
+        engine = Engine.get();
+      }
+      lastTick = Date.now();
+      engine.subscribe('Tick', this, 'onTick');
+    }
+  };
+
+  this.stopMeasuring = function() {
+    if (this.measuring) {
+      this.measuring = false;
+      engine.unsubscribe('Tick', this, 'onTick');
+    }
+  };
+
+  this.addWorstTick = function (delta) {
+    worstTicks.push(delta);
+    worstTicks.sort(function(a, b) {
+      return a - b;
+    });
+
+    if (worstTicks.length > worstTicksSize) {
+      worstTicks.shift();
+    }
+
+    ticksSinceLastWorstUpdate = 0;
+  };
+
+  this.onTick = function(dt) {
+    var now = Date.now();
+    var delta = now - lastTick;
+    var isWorstDelta = (worstTicks.length > 0 && delta >= worstTicks[0]);
+    
+    lastTick = now;
+    
+    if (isWorstDelta || worstTicks.length === 0) {
+      this.addWorstTick(delta);
+    } else {
+      ticksSinceLastWorstUpdate++;
+    }
+
+    if (ticksSinceLastWorstUpdate > TICKS_TIL_INCREASE_SCORE) {
+      worstTicks.pop();
+      ticksSinceLastWorstUpdate = 0;
+    }
+  };
+
+  this.getWorstTicksAverage = function () {
+    var worstTicksAverage = 0;
+    for (var i = 0; i < worstTicks.length; i++) {
+      worstTicksAverage += worstTicks[i];
+    }
+
+    return worstTicksAverage /= worstTicks.length;
+  };
+
+  this.getPerformanceScore = function() {
+    var worstTicksAverage = this.getWorstTicksAverage();
+    var ticksPerSecond = 1000 / worstTicksAverage;
+    var adjustedTicksPerSecond = Math.min(ticksPerSecond, 60);
+    var mappedScore = _map(adjustedTicksPerSecond, minFPS, maxFPS, 0, 100);
+
+    return Math.max(0, mappedScore);
+  };
+
+  this.getAdjustedParticleCount = function(count, performanceRank, allowReduction) {
```

Comment:
I think `performanceScore` is a better name, and is more clear, let's change all instances of `performanceRank` to `performanceScore`